device/core/src/pnp/properties.ts

```diff
@@ -0,0 +1,92 @@
+import { JSONSerializableValue, JSONSerializableObject } from '.';
+import { Twin } from '../twin';
+
+/**
+ * Helper function to generate an object used in the device's response to a writable property request.
+ *
+ * @param {JSONSerializableValue} value          The the value field of the response object. Usually equal to the received value in the writable property request.
+ * @param {number}                ackCode        The ac field of the response object.
+ * @param {string}                ackDescription The ad field of the response object.
+ * @param {number}                ackVersion     The av field of the response object. Usually equal to the received version number in the writable property request.
+ */
+export function generateWritablePropertyResponse(value: JSONSerializableValue, ackCode: number, ackDescription: string, ackVersion: number): {
+  value: JSONSerializableValue,
+  ac: number,
+  ad: string,
+  av: number
+} {
+  return {
+    value,
+    ac: ackCode,
+    ad: ackDescription,
+    av: ackVersion
+  };
+}
+
+export class ClientPropertyCollection {
+  /**
+   * The object representation of the properties.
+   */
+  backingObject: JSONSerializableObject;
+
+  constructor(backingObject?: JSONSerializableObject) {
+    this.backingObject = backingObject ?? {};
+  }
+
+  /**
+   * The $version field of the properties.
+   * Returns NaN if the $version field cannot be converted to a number.
+   * Equivalent to Number(backingObject.$version)
+   */
+  get version(): number {
+    return Number(this.backingObject.$version);
+  }
+
+  /**
+   * Assigns the provided value to a specified property for a component.
+   * If no component is specified, the default component is assumed.
+   *
+   * @param {string}                componentName The name of the component.
+   * @param {string}                propertyName  The name of the property.
+   * @param {JSONSerializableValue} value         The value to assign to the property.
+   */
+  setProperty(propertyName: string, value: JSONSerializableValue): void;
+  setProperty(componentName: string, propertyName: string, value: JSONSerializableValue): void;
+  setProperty(propertyNameOrComponentName: string, valueOrPropertyName: JSONSerializableValue, value?: JSONSerializableValue): void {
+    if (value === undefined) { // We cannot just check the truthiness of value because null is a valid argument for value
```

Comment:
I think you should describe more the structure that you are creating here.  More explanation on why your normalization is correct.

Reply:
Added some explanation in the form of comments